src/libraries/System.IO.Pipes/src/System/IO/Pipes/PipeOptions.cs

```diff
@@ -9,6 +11,8 @@ public enum PipeOptions
         None = 0x0,
         WriteThrough = unchecked((int)0x80000000),
         Asynchronous = unchecked((int)0x40000000),  // corresponds to FILE_FLAG_OVERLAPPED
-        CurrentUserOnly = unchecked((int)0x20000000)
+        CurrentUserOnly = unchecked((int)0x20000000),
+        [SupportedOSPlatform("windows")]
```

Comment:
There's no way to make this work on non-Windows?

Reply:
I think we can make it work on Unix as well, 
Technically pipe is also a file on Unix right ?
So if we can check if a file with that path exists before we go for [creation of named pipe](https://github.com/dotnet/runtime/blob/951984aea84232e290acaee5d98873880c5872c6/src/libraries/System.IO.Pipes/src/System/IO/Pipes/NamedPipeServerStream.Unix.cs#L52) and throw an equivalent exception.
But I'd like to know opinion on this approach ?

Reply:
I took at a look at how it's implemented on Unix. 

The `maxNumberOfServerInstances` is emulated by keeping a static cache, so it works only for the current process:

https://github.com/dotnet/runtime/blob/b883e2d25736432bfe1a6af4896aed2af3d69ef9/src/libraries/System.IO.Pipes/src/System/IO/Pipes/NamedPipeServerStream.Unix.cs#L41-L43

But what is interesting is this comment:

https://github.com/dotnet/runtime/blob/b883e2d25736432bfe1a6af4896aed2af3d69ef9/src/libraries/System.IO.Pipes/src/System/IO/Pipes/NamedPipeServerStream.Unix.cs#L316-L322

So it seems like we could implement `FirstPipeInstance` by not calling the `Unlink` when this flag is set and when the creation of socket fails, map the error to `UnauthorizedAccessException`.

```cs
private SharedServer(string path, int maxCount, PipeOptions pipeOptions)
{
    bool firstPipeInstance = (pipeOptions & PipeOptions.FirstPipeInstance) != 0;

    if (!firstPipeInstance)
    {
        // Binding to an existing path fails, so we need to remove anything left over at this location.
        // There's of course a race condition here, where it could be recreated by someone else between this
        // deletion and the bind below, in which case we'll simply let the bind fail and throw.
        Interop.Sys.Unlink(path); // ignore any failures
    }

    // Start listening for connections on the path.
    var socket = new Socket(AddressFamily.Unix, SocketType.Stream, ProtocolType.Unspecified);
    try
    {
        socket.Bind(new UnixDomainSocketEndPoint(path));
        socket.Listen(int.MaxValue);
    }
    catch ($SomeSpecificException ex) when (firstPipeInstance)
    {
        socket.Dispose();
        throw new UnauthorizedAccessException($someText);
    }
    catch
    {
        socket.Dispose();
        throw;
    }
}
```

Reply:
I've updated the code so that it works on Linux as well, 
Good catch @adamsitnik we can use the conditional Unlinking for making the FirstPipeInstance work across processes
However within the same process we use the static dictionary to check if a server for the given path already exists.
I've changed the code in this area such that it checks if there is firstPipeInstance applied and throw exception.